Fix TLVF for TLVs with a type of two bytes

arnout opened the pull request:
As part of the encryption work, we discovered that the tlvf currently assumes
a fixed TLV "header" of 3 bytes. However, the TLVs used by WSC are actually four
bytes: two bytes type and two bytes value.

This PR fixes that issue by using the actual sizes.

As part of this, a test is added for TLV lists (the issue is triggered for dynamic lists).
However, this test uncovers a number of other issues, so those are fixed as well.

arnout committed:
tlvf: test: add test of TLV lists

There are various ways to build lists in tlvf. Add a test to check if
all these different lists really work.

We already have a tlvTestVarList class that can be used to do this test.

The test currently fails in various ways. Follow-up commits will fix the
failures.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: fix swapping of simple list of scalars

The swapping function for the entries of a list of scalars was missing
an address-of.

This issue was not detected before because the only simple lists that
are actually used have uint8_t members, which don't need swapping. All
other lists have class members which have a swap function so they don't
need an ampersand.

Also, the compiler doesn't detect this because we use reinterpret_cast.

Fix this by adding the address-of. We also add it for classes, even
though it's never actually added there because classes have swap_is_func
== True.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented:
small thing about the TLVF potential issues -
1. Allocation of lists have to be in the same order they are defined in the yaml. This can be optimized (I thought of a better algorithm) but no point in putting effort on this currently. We should open a task on this.
2. variable length lists of classes - create - alloc - add have to be in order. Changing variables which come after a list **has** to be after calling add since the pointers are "dangling" before hand.
3. Other bugs you found - please list them somewhere so I know what to expect and perhaps seen them and thought they are not worth the effort.

tomereli commented:
Perhaps I'm missing something or it hasn't been pushed yet, but I didn't see the fix to the issue we root caused - fixing kMinimumLength from 3 to 4 (actually using the types to determine the size)

arnout committed:
tlvf: test: output to log instead of std::cout

There were a few debug lines left that were using std::cout for output.
This output is still useful, but it should go to the log file instead.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: tlvTestVarList: correct position of length field

The tlvf doesn't correctly handle TLV types where the length field is
not the second one. Since we don't need that anyway, just fix the
tlvfTestVarList definition to have the length as the second field
instead of the third field.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: tlvfTestVarList: use 1 byte for list lengths

The tlvf currently doesn't support a list length of more than one byte,
because it handles swapping incorrectly. Since we don't actually need
list lengths with more than one byte, it will not be fixed now.
Therefore, just change tlvfTestVarList do always use one byte for the
list length.

The problem lies in the class_swap function: it will first swap the
length member and then use this length member to iterate over the
elements. For parsing, this is indeed correct. For sending, however,
after swapping the length would no longer be correct, so the length has
to be swapped after iterating over the list instead of before. For
1-byte length this is not a problem since it doesn't need to be swapped.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: test: change types after change of tlvTestVarList lengths

Since the length fields changed types, also update the types in the
test.

uint8_t is printed as ASCII by default, so convert to unsigned for
proper output.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: test: disable complex_list test

The complex list doesn't work correctly if it has more than one element.

For now we don't need it, so disable this test for the moment. The issue
is tracked in #137.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: support type of two (or more) bytes

Up to now, we only had TLV types where the type is 1 byte. However, the
WSC TLVs have a type field of two bytes. It turns out that the tlvf
doesn't support this.

The problem is the tlvf calculates the dynamic length of an element by
subtracting the pointer to the beginning of the field from the pointer
to the end of the struct. But to calculate the end pointer, it needs to
know the length, and the length starts *after* the length field. So the
size of the type and length fields has to be subtracted. Up to now, this
was done with constant 3 (kMinimumLength): 1 byte type and 2 bytes
length.

To fix this, use the actual sizeof the two fields. We could generate the
exact size based on the type information in tlvf, but we don't have that
readily available, so sizeof is much simpler.

Add a test to validate this, by making the type field of tlvTestVarLists
two bytes instead of one.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tlvf: regenerate after previous commit

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on framework/tlvf/test/main.cpp:
```diff
@@ -152,6 +153,49 @@ int main(int argc, char *argv[])
     LOG(DEBUG) << "Done (WSC M2)";
     MAPF_INFO("TLV LENGTH WSC M2: " << thirdTlv->length());
 
+    auto fourthTlv     = msg.addClass<tlvTestVarList>();
```
Consider putting each test in each own function - so there will be a TX method and RX method, will make `main` more readable and extensible

tomereli commented on framework/tlvf/test/main.cpp:
```diff
@@ -152,6 +153,49 @@ int main(int argc, char *argv[])
+    auto fourthTlv     = msg.addClass<tlvTestVarList>();
+    fourthTlv->var0()  = 0xa0;
+    allocation_succeed = fourthTlv->alloc_simple_list(2);
+    mapf_assert(allocation_succeed);
```
Perhaps it would be better to align all tests to the way you wrote the encryption_test, with the `check` function.
Using assert for some failures and incrementing errors on others is inconsistent.

arnout replied:
Agreed, but it's a copy from earlier.

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -285,12 +285,12 @@ bool tlvTestVarList::init()
+    m_var0 = (uint16_t*)m_buff_ptr__;
+    m_buff_ptr__ += sizeof(uint16_t) * 1;
+    if(m_length && !m_parse__){ (*m_length) += sizeof(uint16_t); }
     m_simple_list_length = (uint16_t*)m_buff_ptr__;
```
I think this commit can be squashed to the previous one (nitpick)
In general I think that splitting to autogenerated commits is worth it only if there are a lot of autogenerated changes.

arnout replied:
Note that github doesn't keep track of which commit was open when you type a review comment. So "this commit" is fairly meaningless.

I agree that splitting off autogenerated commits is not very valuable. Conversely, I found it very useful when reviewing #139 to have the tlvf.py change together with the source code changes, because you immediately see the effect.

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/test/tlvVarList.h:
```diff
@@ -33,10 +33,10 @@ class tlvTestVarList : public BaseClass
         const uint8_t& type();
         const uint16_t& length();
         uint16_t& var0();
-        uint16_t& simple_list_length();
```
Why 2 autogenerated commits one after the other? looks strange, the previous one did not change tlvf.py. Am I missing something?

arnout replied:
That's because github doesn't display commits in the correct order. They are not in the actual order, but instead they are ordered by date. For this particular commit, I accidentally copied the timestamp of the previous autogenerated commit (by using `-c HEAD^` because I was too lazy to type the same commit message again). So the commit date of this commit is before the commit date of the change to tlvVarList that triggers it.

Anyway, I'm going to rebase and squash, like you proposed, so the problem will go away :-)

tomereli commented on framework/tlvf/test/main.cpp:
```diff
@@ -171,16 +171,18 @@ int main(int argc, char *argv[])
-        LOG(ERROR) << "Could add complex list a second time";
-        errors++;
-    }
+    // TODO the complex list doesn't work at the moment if it has more than one element
```
As I suggested in previous commit - please take this out to a function to keep main simple to read

tomereli reviewed (approved):
Nothing blocking - refactoring the tlvf test to be more readable can be done in a different PR. Please open a task for it in case you choose not to fix it as part of this PR/

arnout commented:
> refactoring the tlvf test to be more readable can be done in a different PR.
> Please open a task for it in case you choose not to fix it as part of this PR.

I don't think it's worth opening a separate task or PR for refactoring a test. It's something that can be done as part of updating the test for some other reason.
